### PR TITLE
Add operator selection to Mobile Connect authentication flow

### PR DESCRIPTION
**Pull request by gorandostanic**
In order to enable Mobile Connect authentication on devices without SIM card, we divert auth flow to operator selection. The actual initialisation is delayed until the interception of the redirect URI.

### CODE REVIEW

File: connect/src/com/telenor/connect/SdkProfile.java [hunk 1]
```diff
@@ -24,10 +24,23 @@
...
+            Map<String, String> parameters,
+            List<String> uiLocales,
+            OnStartAuthorizationCallback callback);
+
+    interface OnDeliverAuthorizationUriCallback {
+        void onSuccess(Uri authorizationStartUri);
         void onError();
     }
+
+    interface OnStartAuthorizationCallback extends OnDeliverAuthorizationUriCallback {
```
Comment (phyrex1an) [inline connect/src/com/telenor/connect/SdkProfile.java]: Any reason why having this is the same hierarchy? Why isn't there a completely separate callback for mobile connect for the operator discovery result (irregardless of the result was delivered via the manual selection or automatically)? It looks weird to have both kinds of callbacks in the same class (eg, sometimes you implement the onDivert method by throwing RuntimeException).

File: connect/src/com/telenor/mobileconnect/MobileConnectSdkProfile.java [hunk 2]
```diff
@@ -100,32 +103,123 @@ public Uri getAuthorizeUri(Map<String, String> parameters, List<String> locales)
     }
 
     @Override
-    public void onStartAuthorization(final OnStartAuthorizationCallback callback) {
+    public synchronized void onStartAuthorization(
```
Comment (phyrex1an) [inline connect/src/com/telenor/mobileconnect/MobileConnectSdkProfile.java]: I really dislike these synchronized methods, including the already present ones. The scope is way to wide, and it's not at all clear what is the actions which needs serialization. Can you rewrite into using explicit locks instead, with as small scope as possible? Alternatively get rid of the volatile usage in favor of explicit state passing between threads, but I think we already discussed that and said it was infeasible in these patches...

File: connect/src/com/telenor/connect/ui/ConnectWebViewClient.java [hunk 3]
```diff
@@ -87,11 +88,47 @@ public ConnectWebViewClient(
     }
 
     @Override
-    public boolean shouldOverrideUrlLoading(WebView view, String url) {
-        if (ConnectSdk.getRedirectUri() != null
-                && url.startsWith(ConnectSdk.getRedirectUri())) {
-            ConnectUtils.parseAuthCode(url, connectCallback);
-            return true;
+    @SuppressWarnings("unchecked")
+    public boolean shouldOverrideUrlLoading(final WebView view, final String url) {
```
Comment (phyrex1an) [inline connect/src/com/telenor/connect/ui/ConnectWebViewClient.java]: This method looks quite misnamed now.

File: connect/src/com/telenor/connect/SdkProfile.java (hunk 1 above)
Comment (atanasdb) [inline connect/src/com/telenor/connect/SdkProfile.java]: I agree with that.
Also for me Operator Selection should be just a different activity, not part of the ConnectActivity.
And you should not modify ConnectWebViewClient.java with the code for Operator Selection.
The flow should be:
getMccMnc -> yes -> getWellKnown -> startAuthroization(wellKnown)
getMccMnc -> no -> getOSUri -> getOperator -> getWellKnown -> startAuthroization(wellKnown)

Comment (atanasdb) [inline connect/src/com/telenor/connect/SdkProfile.java]: Why does my comment goes after Goran's even though ts are correct :)?
"I agree with that" is to the @phyrex1an's comment.

File: connect/src/com/telenor/connect/ui/ConnectWebViewClient.java (hunk 3 above)
Comment (atanasdb) [inline connect/src/com/telenor/connect/ui/ConnectWebViewClient.java]: I think, the name confusion comes form the fact you are placing code where it does not really belong. See me other comment.

File: connect/src/com/telenor/connect/SdkProfile.java (hunk 1 above)
Comment (gorandostanic) [inline connect/src/com/telenor/connect/SdkProfile.java]: I kind of disagree for a couple of reasons listed below (ideally, operator selection should have been a part of a normal OAuth2 flow, in which case we wouldn't have had discussions of this kind in the first place). 

1. A separate OperatorSelection activity would require publishing it in the manifest, thus exposing developers to an activity they are not supposed to use. I think they generally benefit from encapsulations of this kind (the less - the better).

2. ConnectWebViewClient is not modified with OperatorSelection code, but with a more generic thing that the MobileConnect uses for operator selection. Connect Id could have used it for something else, provided they decided to implement preamble diversions. The fact that they don't do it now does not mean they couldn't.

Comment (atanasdb) [inline connect/src/com/telenor/connect/SdkProfile.java]: 1. I'm not sure I understand this argument. Putting an activity in a manifest, shares this with the OS, and does not expose developers to it more than it is currently. Also the encapsulation you did, makes the code a bit harder to follow.
2. "Connect Id could have used it for something else" like what? I think this shouldOverrideUrlLoading is the wrong hook to achieve what you want.

**Comment by gorandostanic** [comment]
1. What I meant was that it's developers that publish activities in the
manifest, so I think it's generally detrimental that they have to publish
activities they are not supposed to use themselves, merely to keep the SDK
in shape.
2. I am not sure it's a wrong hook... my understanding is that the hook is
there to answer "yes" if you don't want the url to be rendered by the
WebView, giving the application a chance to do something else instead.

File: connect/src/com/telenor/connect/SdkProfile.java (hunk 1 above)
Comment (gorandostanic) [inline connect/src/com/telenor/connect/SdkProfile.java]: 1. What I meant was that it's developers that publish activities in the
manifest, so I think it's generally detrimental that they have to publish
activities they are not supposed to use themselves, merely to keep the SDK
in shape.
2. I am not sure it's a wrong hook... my understanding is that the hook is
there to answer "yes" if you don't want the url to be rendered by the
WebView, giving the application a chance to do something else instead.

Comment (atanasdb) [inline connect/src/com/telenor/connect/SdkProfile.java]: 1. They don't have to publish anything if you move the activity into the SDK manifest.

2.  You are putting something that's very generally used in a very low level implementation hook. E.g. let's say they don't use WebViews, but something else? Do you still want your OperatorSelection handling to live there.
2b smells like YAGNI, KISS pls :)

Comment (atanasdb) [inline connect/src/com/telenor/connect/SdkProfile.java]: And "something else" from 2. is quite real: they have plans to drop the WebView.

Comment (gorandostanic) [inline connect/src/com/telenor/connect/SdkProfile.java]: OK, you have convinced me, I'll do a rework :)

Comment (atanasdb) [inline connect/src/com/telenor/connect/SdkProfile.java]: yay

File: connect/src/com/telenor/connect/ConnectSdk.java [hunk 4]
```diff
@@ -146,22 +135,59 @@ public static synchronized void authenticate(final Activity activity,
...
-            public void onError() {
-                showAuthCancelMessage(activity);
-            }
+                    @Override
+                    public void onSuccess(Uri authorizationStartUri) {
+                        startAuthActivity(authorizationStartUri, false);
+                    }
+
+                    @Override
+                    public void onDivert(Uri diversionUri) {
```
Comment (phyrex1an) [inline connect/src/com/telenor/connect/ConnectSdk.java]: I'm guessing the work done per https://github.com/telenordigital/connect-android-sdk/pull/54/files/8514c618036b23ada24c6ba55e45b03c3ad6c0b2#diff-edfffb4e70b7f3f2268239f7a63b2636R43 moves this method somewhere else? If that's the case, consider if the naming could be made less generic.

**Pull request closed without merging**